File: app/src/main/java/de/mchllngr/quickopen/receiver/PackageFullyRemovedReceiver.java
package de.mchllngr.quickopen.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.Nullable;
import de.mchllngr.quickopen.model.ApplicationModel;
import de.mchllngr.quickopen.service.NotificationService;
import timber.log.Timber;

/**
 * {@link BroadcastReceiver} for removing apps from the list when uninstalled.
 */
public class PackageFullyRemovedReceiver extends BroadcastReceiver {

    /**
     * Checks if  the {@link NotificationService} if called with the correct actions.
     */
    @Override
    public void onReceive(Context context, Intent intent) {
        if (context == null || intent == null) return;

        String action = intent.getAction();
        if (isCorrectAction(action)) {
            Timber.d("Action '" + action + "' received");
            ApplicationModel.removeNotLaunchableAppsFromList(context);
        } else
            Timber.w("Unknown action '" + action + "' received");
    }

    /**
     * Returns true when the received action is correct for this {@link BroadcastReceiver}.
     *
     * @param action received action
     * @return true is the action is correct for this {@link BroadcastReceiver}
     */
    private boolean isCorrectAction(@Nullable String action) {
        return Intent.ACTION_PACKAGE_FULLY_REMOVED.equals(action);
    }
}

File: app/src/main/java/de/mchllngr/quickero/receiver/PackageFullyRemovedReceiver.kt
package de.mchllngr.quickero.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import dagger.hilt.android.AndroidEntryPoint
import de.mchllngr.quickero.repository.application.ApplicationsRepository
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/** [BroadcastReceiver] for removing apps from the list when uninstalled.*/
@AndroidEntryPoint
class PackageFullyRemovedReceiver : BroadcastReceiver() {

    @Inject lateinit var applicationsRepository: ApplicationsRepository

    override fun onReceive(
        context: Context?,
        intent: Intent?
    ) {
        if (context == null || intent == null) return

        val action = intent.action
        if (Intent.ACTION_PACKAGE_FULLY_REMOVED == action) {
            Timber.d("Action '$action' received")
            GlobalScope.launch {
                applicationsRepository.removeInvalidApplications()
            }
        } else {
            Timber.w("Unknown action '$action' received")
        }
    }
}
